File: sandbox-ts/protocol/12020-tcp/equipment.ts
/**
 * Equipment Protocol Packets
 *
 * Equipment, weapons, relics, and related structures.
 */

import type { BinaryReader, BinaryWriter } from "./binary";
import { type CCommonRespMsg, writeCCommonRespMsg } from "./common";

// =============================================================================
// EQUIPMENT REQUESTS
// =============================================================================

export interface CReqEquipWear {
  m_nTransID: number; // UInt32
  m_nType: number; // UInt16
  m_nEquipUniqueId: bigint; // UInt64
  m_nSlotId: number; // UInt16
}

export function readCReqEquipWear(reader: BinaryReader): CReqEquipWear {
  return {
    m_nTransID: reader.readUInt32(),
    m_nType: reader.readUInt16(),
    m_nEquipUniqueId: reader.readUInt64(),
    m_nSlotId: reader.readUInt16(),
  };
}

export interface CReqEquipTotem {
  m_nTransID: number; // UInt32
  m_nType: number; // UInt16
  m_nTotemId: number; // UInt32
}

export function readCReqEquipTotem(reader: BinaryReader): CReqEquipTotem {
  return {
    m_nTransID: reader.readUInt32(),
    m_nType: reader.readUInt16(),
    m_nTotemId: reader.readUInt32(),
  };
}

export interface CEquipRefine {
  m_nType: number; // UInt16
  m_nTransID: number; // UInt32
  m_nPosId: number; // UInt16
  m_nCarvingId: number; // UInt32
  m_nCarvingIdx: number; // UInt16
  arrayEquipId: bigint[]; // UInt64[]
  vecCompositeId: number[]; // UInt32[]
}

export function readCEquipRefine(reader: BinaryReader): CEquipRefine {
  return {
    m_nType: reader.readUInt16(),
    m_nTransID: reader.readUInt32(),
    m_nPosId: reader.readUInt16(),
    m_nCarvingId: reader.readUInt32(),
    m_nCarvingIdx: reader.readUInt16(),
    arrayEquipId: reader.readArray(() => reader.readUInt64()),
    vecCompositeId: reader.readArray(() => reader.readUInt32()),
  };
}

// =============================================================================
// EQUIPMENT RESPONSES
// =============================================================================

export interface CRespEquipWear {
  m_stRetMsg: CCommonRespMsg;
  m_nEquipUniqueId: bigint; // UInt64
  m_nSlotId: number; // UInt16
}

export function writeCRespEquipWear(writer: BinaryWriter, resp: CRespEquipWear): void {
  writeCCommonRespMsg(writer, resp.m_stRetMsg);
  writer.writeUInt64(resp.m_nEquipUniqueId);
  writer.writeUInt16(resp.m_nSlotId);
}

export interface CRespEquipTotem {
  m_stRetMsg: CCommonRespMsg;
  m_nTotemId: number; // UInt32
}

export function writeCRespEquipTotem(writer: BinaryWriter, resp: CRespEquipTotem): void {
  writeCCommonRespMsg(writer, resp.m_stRetMsg);
  writer.writeUInt32(resp.m_nTotemId);
}

export interface CRespEquipRefine {
  m_stRetMsg: CCommonRespMsg;
  m_nResult: number; // UInt16
}

export function writeCRespEquipRefine(writer: BinaryWriter, resp: CRespEquipRefine): void {
  writeCCommonRespMsg(writer, resp.m_stRetMsg);
  writer.writeUInt16(resp.m_nResult);
}

// =============================================================================
// HERO SKIN
// =============================================================================

export interface CReqHeroSkin {
  m_nType: number; // UInt16
  m_nTransID: number; // UInt32
  m_nSkinId: number; // UInt32
  m_nNum: number; // UInt32
}

export function readCReqHeroSkin(reader: BinaryReader): CReqHeroSkin {
  return {
    m_nType: reader.readUInt16(),
    m_nTransID: reader.readUInt32(),
    m_nSkinId: reader.readUInt32(),
    m_nNum: reader.readUInt32(),
  };
}

export interface CHeroSkin {
  m_nSkinId: number; // UInt32
  m_nLevel: number; // UInt16
  m_bIsOwned: boolean; // Boolean
}

export function writeCHeroSkin(writer: BinaryWriter, skin: CHeroSkin): void {
  writer.writeUInt32(skin.m_nSkinId);
  writer.writeUInt16(skin.m_nLevel);
  writer.writeBool(skin.m_bIsOwned);
}

export interface CRespHeroSkin {
  m_stRetMsg: CCommonRespMsg;
  m_vecSkins: CHeroSkin[];
}

export function writeCRespHeroSkin(writer: BinaryWriter, resp: CRespHeroSkin): void {
  writeCCommonRespMsg(writer, resp.m_stRetMsg);
  writer.writeArray(resp.m_vecSkins, (s) => writeCHeroSkin(writer, s));
}

// =============================================================================
// WEAPON SKIN
// =============================================================================

export interface CReqWeaponSkin {
  m_nType: number; // UInt16
  m_nTransID: number; // UInt32
  m_nSkinId: number; // UInt32
}

export function readCReqWeaponSkin(reader: BinaryReader): CReqWeaponSkin {
  return {
    m_nType: reader.readUInt16(),
    m_nTransID: reader.readUInt32(),
    m_nSkinId: reader.readUInt32(),
  };
}

export interface CRespWeaponSkin {
  m_stRetMsg: CCommonRespMsg;
  m_nSkinId: number; // UInt32
}

export function writeCRespWeaponSkin(writer: BinaryWriter, resp: CRespWeaponSkin): void {
  writeCCommonRespMsg(writer, resp.m_stRetMsg);
  writer.writeUInt32(resp.m_nSkinId);
}

// =============================================================================
// WING
// =============================================================================

export interface CReqWing {
  m_nType: number; // UInt16
  m_nTransID: number; // UInt32
  m_nWingId: number; // UInt32
}

export function readCReqWing(reader: BinaryReader): CReqWing {
  return {
    m_nType: reader.readUInt16(),
    m_nTransID: reader.readUInt32(),
    m_nWingId: reader.readUInt32(),
  };
}

export interface CRespWing {
  m_stRetMsg: CCommonRespMsg;
  m_nWingId: number; // UInt32
}

export function writeCRespWing(writer: BinaryWriter, resp: CRespWing): void {
  writeCCommonRespMsg(writer, resp.m_stRetMsg);
  writer.writeUInt32(resp.m_nWingId);
}

// =============================================================================
// BOXES
// =============================================================================

export interface CReqOpenDragonBox {
  m_nType: number; // UInt16
  m_nTransID: number; // UInt32
  m_nDiamond: number; // UInt16
  m_nBatchCount: number; // UInt16
}

export function readCReqOpenDragonBox(reader: BinaryReader): CReqOpenDragonBox {
  return {
    m_nType: reader.readUInt16(),
    m_nTransID: reader.readUInt32(),
    m_nDiamond: reader.readUInt16(),
    m_nBatchCount: reader.readUInt16(),
  };
}

export interface CRespOpenDragonBox {
  m_stRetMsg: CCommonRespMsg;
  m_nBoxCount: number; // UInt16
  m_vecRewards: number[]; // UInt32[]
}

export function writeCRespOpenDragonBox(writer: BinaryWriter, resp: CRespOpenDragonBox): void {
  writeCCommonRespMsg(writer, resp.m_stRetMsg);
  writer.writeUInt16(resp.m_nBoxCount);
  writer.writeArray(resp.m_vecRewards, (r) => writer.writeUInt32(r));
}

export interface CReqOpenPetBox {
  m_nType: number; // UInt16
  m_nTransID: number; // UInt32
  m_nCount: number; // UInt16
}

export function readCReqOpenPetBox(reader: BinaryReader): CReqOpenPetBox {
  return {
    m_nType: reader.readUInt16(),
    m_nTransID: reader.readUInt32(),
    m_nCount: reader.readUInt16(),
  };
}

export interface CRespOpenPetBox {
  m_stRetMsg: CCommonRespMsg;
  m_vecRewards: number[]; // UInt32[]
}

export function writeCRespOpenPetBox(writer: BinaryWriter, resp: CRespOpenPetBox): void {
  writeCCommonRespMsg(writer, resp.m_stRetMsg);
  writer.writeArray(resp.m_vecRewards, (r) => writer.writeUInt32(r));
}

export interface CReqOpenEquipSBox {
  m_nType: number; // UInt16
  m_nTransID: number; // UInt32
  m_nCount: number; // UInt16
}

export function readCReqOpenEquipSBox(reader: BinaryReader): CReqOpenEquipSBox {
  return {
    m_nType: reader.readUInt16(),
    m_nTransID: reader.readUInt32(),
    m_nCount: reader.readUInt16(),
  };
}

export interface CRespOpenEquipSBox {
  m_stRetMsg: CCommonRespMsg;
  m_vecRewards: number[]; // UInt32[]
}

export function writeCRespOpenEquipSBox(writer: BinaryWriter, resp: CRespOpenEquipSBox): void {
  writeCCommonRespMsg(writer, resp.m_stRetMsg);
  writer.writeArray(resp.m_vecRewards, (r) => writer.writeUInt32(r));
}

File: server-nim/protocol/equipment.nim
## Equipment Protocol Packets
##
## Equipment, weapons, relics, and related structures.

import binary, common

# =============================================================================
# EQUIPMENT REQUESTS
# =============================================================================

type
  CReqEquipWear* = object
    m_nTransID*: uint32
    m_nType*: uint16
    m_nEquipUniqueId*: uint64
    m_nSlotId*: uint16

proc readCReqEquipWear*(reader: var BinaryReader): CReqEquipWear =
  result.m_nTransID = reader.readUint32()
  result.m_nType = reader.readUint16()
  result.m_nEquipUniqueId = reader.readUint64()
  result.m_nSlotId = reader.readUint16()

type
  CReqEquipTotem* = object
    m_nTransID*: uint32
    m_nType*: uint16
    m_nTotemId*: uint32

proc readCReqEquipTotem*(reader: var BinaryReader): CReqEquipTotem =
  result.m_nTransID = reader.readUint32()
  result.m_nType = reader.readUint16()
  result.m_nTotemId = reader.readUint32()

type
  CEquipRefine* = object
    m_nType*: uint16
    m_nTransID*: uint32
    m_nPosId*: uint16
    m_nCarvingId*: uint32
    m_nCarvingIdx*: uint16
    arrayEquipId*: seq[uint64]
    vecCompositeId*: seq[uint32]

proc readCEquipRefine*(reader: var BinaryReader): CEquipRefine =
  result.m_nType = reader.readUint16()
  result.m_nTransID = reader.readUint32()
  result.m_nPosId = reader.readUint16()
  result.m_nCarvingId = reader.readUint32()
  result.m_nCarvingIdx = reader.readUint16()
  result.arrayEquipId = reader.readArray(proc(r: var BinaryReader): uint64 = r.readUint64())
  result.vecCompositeId = reader.readArray(proc(r: var BinaryReader): uint32 = r.readUint32())

# =============================================================================
# EQUIPMENT RESPONSES
# =============================================================================

type
  CRespEquipWear* = object
    m_stRetMsg*: CCommonRespMsg
    m_nEquipUniqueId*: uint64
    m_nSlotId*: uint16

proc writeCRespEquipWear*(writer: var BinaryWriter, resp: CRespEquipWear) =
  writer.writeCCommonRespMsg(resp.m_stRetMsg)
  writer.writeUint64(resp.m_nEquipUniqueId)
  writer.writeUint16(resp.m_nSlotId)

type
  CRespEquipTotem* = object
    m_stRetMsg*: CCommonRespMsg
    m_nTotemId*: uint32

proc writeCRespEquipTotem*(writer: var BinaryWriter, resp: CRespEquipTotem) =
  writer.writeCCommonRespMsg(resp.m_stRetMsg)
  writer.writeUint32(resp.m_nTotemId)

type
  CRespEquipRefine* = object
    m_stRetMsg*: CCommonRespMsg
    m_nResult*: uint16

proc writeCRespEquipRefine*(writer: var BinaryWriter, resp: CRespEquipRefine) =
  writer.writeCCommonRespMsg(resp.m_stRetMsg)
  writer.writeUint16(resp.m_nResult)

# =============================================================================
# HERO SKIN
# =============================================================================

type
  CReqHeroSkin* = object
    m_nType*: uint16
    m_nTransID*: uint32
    m_nSkinId*: uint32
    m_nNum*: uint32

proc readCReqHeroSkin*(reader: var BinaryReader): CReqHeroSkin =
  result.m_nType = reader.readUint16()
  result.m_nTransID = reader.readUint32()
  result.m_nSkinId = reader.readUint32()
  result.m_nNum = reader.readUint32()

type
  CHeroSkin* = object
    m_nSkinId*: uint32
    m_nLevel*: uint16
    m_bIsOwned*: bool

proc writeCHeroSkin*(writer: var BinaryWriter, skin: CHeroSkin) =
  writer.writeUint32(skin.m_nSkinId)
  writer.writeUint16(skin.m_nLevel)
  writer.writeBool(skin.m_bIsOwned)

type
  CRespHeroSkin* = object
    m_stRetMsg*: CCommonRespMsg
    m_vecSkins*: seq[CHeroSkin]

proc writeCRespHeroSkin*(writer: var BinaryWriter, resp: CRespHeroSkin) =
  writer.writeCCommonRespMsg(resp.m_stRetMsg)
  writer.writeArray(resp.m_vecSkins, proc(w: var BinaryWriter, s: CHeroSkin) = w.writeCHeroSkin(s))

# =============================================================================
# WEAPON SKIN
# =============================================================================

type
  CReqWeaponSkin* = object
    m_nType*: uint16
    m_nTransID*: uint32
    m_nSkinId*: uint32

proc readCReqWeaponSkin*(reader: var BinaryReader): CReqWeaponSkin =
  result.m_nType = reader.readUint16()
  result.m_nTransID = reader.readUint32()
  result.m_nSkinId = reader.readUint32()

type
  CRespWeaponSkin* = object
    m_stRetMsg*: CCommonRespMsg
    m_nSkinId*: uint32

proc writeCRespWeaponSkin*(writer: var BinaryWriter, resp: CRespWeaponSkin) =
  writer.writeCCommonRespMsg(resp.m_stRetMsg)
  writer.writeUint32(resp.m_nSkinId)

# =============================================================================
# WING
# =============================================================================

type
  CReqWing* = object
    m_nType*: uint16
    m_nTransID*: uint32
    m_nWingId*: uint32

proc readCReqWing*(reader: var BinaryReader): CReqWing =
  result.m_nType = reader.readUint16()
  result.m_nTransID = reader.readUint32()
  result.m_nWingId = reader.readUint32()

type
  CRespWing* = object
    m_stRetMsg*: CCommonRespMsg
    m_nWingId*: uint32

proc writeCRespWing*(writer: var BinaryWriter, resp: CRespWing) =
  writer.writeCCommonRespMsg(resp.m_stRetMsg)
  writer.writeUint32(resp.m_nWingId)

# =============================================================================
# BOXES
# =============================================================================

type
  CReqOpenDragonBox* = object
    m_nType*: uint16
    m_nTransID*: uint32
    m_nDiamond*: uint16
    m_nBatchCount*: uint16

proc readCReqOpenDragonBox*(reader: var BinaryReader): CReqOpenDragonBox =
  result.m_nType = reader.readUint16()
  result.m_nTransID = reader.readUint32()
  result.m_nDiamond = reader.readUint16()
  result.m_nBatchCount = reader.readUint16()

type
  CRespOpenDragonBox* = object
    m_stRetMsg*: CCommonRespMsg
    m_nBoxCount*: uint16
    m_vecRewards*: seq[uint32]

proc writeCRespOpenDragonBox*(writer: var BinaryWriter, resp: CRespOpenDragonBox) =
  writer.writeCCommonRespMsg(resp.m_stRetMsg)
  writer.writeUint16(resp.m_nBoxCount)
  writer.writeArray(resp.m_vecRewards, proc(w: var BinaryWriter, r: uint32) = w.writeUint32(r))

type
  CReqOpenPetBox* = object
    m_nType*: uint16
    m_nTransID*: uint32
    m_nCount*: uint16

proc readCReqOpenPetBox*(reader: var BinaryReader): CReqOpenPetBox =
  result.m_nType = reader.readUint16()
  result.m_nTransID = reader.readUint32()
  result.m_nCount = reader.readUint16()

type
  CRespOpenPetBox* = object
    m_stRetMsg*: CCommonRespMsg
    m_vecRewards*: seq[uint32]

proc writeCRespOpenPetBox*(writer: var BinaryWriter, resp: CRespOpenPetBox) =
  writer.writeCCommonRespMsg(resp.m_stRetMsg)
  writer.writeArray(resp.m_vecRewards, proc(w: var BinaryWriter, r: uint32) = w.writeUint32(r))

type
  CReqOpenEquipSBox* = object
    m_nType*: uint16
    m_nTransID*: uint32
    m_nCount*: uint16

proc readCReqOpenEquipSBox*(reader: var BinaryReader): CReqOpenEquipSBox =
  result.m_nType = reader.readUint16()
  result.m_nTransID = reader.readUint32()
  result.m_nCount = reader.readUint16()

type
  CRespOpenEquipSBox* = object
    m_stRetMsg*: CCommonRespMsg
    m_vecRewards*: seq[uint32]

proc writeCRespOpenEquipSBox*(writer: var BinaryWriter, resp: CRespOpenEquipSBox) =
  writer.writeCCommonRespMsg(resp.m_stRetMsg)
  writer.writeArray(resp.m_vecRewards, proc(w: var BinaryWriter, r: uint32) = w.writeUint32(r))
